server/models.py

```diff
@@ -254,43 +247,136 @@ class GroupMember(db.Model, TimestampMixin):
-        Left - The user accept the invite but later left.
-        Removed - The user was removed by another meber.
-    This distinction is neccesary because a user may leave multiple groups.
+    A user may only be invited or participate in a single group per assignment.
+
+    The status value can be one of:
+        pending - The user has been invited to the group.
+        active  - The user accepted the invite and is part of the group.
     """
-    status_values = ['invited', 'accepted']
-    detail_values = ['revoked', 'declined', 'left', 'removed']
+    __tablename__ = 'GroupMember'
+    __table_args__ = (
+        PrimaryKeyConstraint('user_id', 'assignment_id', name='pk_GroupMember'),
+    )
+    status_values = ['pending', 'active']
 
-    id = db.Column(db.Integer(), primary_key=True)
-    assignment = db.Column(db.ForeignKey("assignment.id"), nullable=False)
-    group = db.Column(db.ForeignKey("group.id"), nullable=False, index=True)
-    user = db.Column(db.ForeignKey("user.id"), nullable=False, index=True)
-    extra = db.Column(db.String())  # e.g. Member "A"
+    user_id = db.Column(db.ForeignKey("user.id"), nullable=False, index=True)
+    assignment_id = db.Column(db.ForeignKey("assignment.id"), nullable=False)
+    group_id = db.Column(db.ForeignKey("group.id"), nullable=False, index=True)
 
     status = db.Column(db.Enum(*status_values, name="status"), index=True)
-    detail = db.Column(db.Enum(*detail_values, name='detail'))
     updated = db.Column(db.DateTime, onupdate=db.func.now())
 
-    UniqueConstraint('assignment', 'user', 'status', name='uq_userInOneGroup')
+    user = db.relationship("User")
+    assignment = db.relationship("Assignment")
+    group = db.relationship("Group",
+        backref=backref('members', cascade="all, delete-orphan"))
 
 
 class Group(db.Model, TimestampMixin):
     """A group is a collection of users who are either members or invited.
 
-    Members of a group can view each other's submissions.
+    Groups are created when a member not in a group invites another member.
+    Invited members may accept or decline invitations. Active members may
+    revoke invitations and remove members (including themselves).
 
-    Specification:
-    https://github.com/Cal-CS-61A-Staff/ok/wiki/Group-&-Submission-Consistency
+    A group must have at least 1 active member and at lease 2 participants.
+    Degenerates groups are deleted.
     """
     id = db.Column(db.Integer(), primary_key=True)
-    assignment = db.Column(db.ForeignKey("assignment.id"), nullable=False)
-    members = db.relationship("GroupMember")
+    assignment_id = db.Column(db.ForeignKey("assignment.id"), nullable=False)
+
+    assignment = db.relationship("Assignment")
+
+    def size(self):
+        return GroupMember.query.filter_by(group=self).count()
+
+    def has_status(self, user, status):
+        return GroupMember.query.filter_by(
+            user=user,
+            group=self,
```

Comment:
I'm not sure what the ORM does here since group is a relationship. I can find out in a bit. (mostly just curious about what the ORM is doing) 

The alternative is to do `group_id=self.id`